### PR TITLE
IC-13 feat: implement LearnWorlds transform and persistence pipeline 

### LINKED ISSUE #82: IC-13 feat: implement LearnWorlds data validation, transformation, and persistence pipeline
### Title
IC-13 feat: implement LearnWorlds data validation, transformation, and persistence pipeline

---

## Description
Implement the application-layer pipeline that validates raw LearnWorlds records, transforms them into the internal canonical model, persists valid records, skips invalid records, and provides retrieval by learner identifier. This issue explicitly excludes database schema work.

---

## Goals
- Validate and normalize incoming LearnWorlds records  
- Persist valid records using idempotent update strategy  
- Skip malformed records without stopping the batch  
- Produce processing summaries for observability  
- Provide reliable retrieval by learner identifier  

---

### Validation Rules
- Required field presence checks  
- Type checks and coercion where safe  
- Timestamp normalization  
- Progress bounds enforcement  

### Transformation Rules
- Map raw LearnWorlds fields into canonical internal structure  
- Normalize completion status values  
- Compute completed modules where applicable  

### Persistence Behavior
- Upsert strategy for repeated ingestion runs  
- Batch processing that continues on per-record failures  

### Summary Reporting
- Total processed  
- Valid persisted  
- Invalid skipped  

### Retrieval Behavior
- Query and return consistent learner progress by learner identifier  

---

## Non-Functional Requirements
- **Idempotent processing** across reruns  
- **Deterministic transformation** behavior  
- **Clear, structured logging** with no secret leakage  
- **Strict separation** between raw external payload handling and normalized internal model  

---

### Scenario 1: Mixed Input Handling
**Given** mixed valid and invalid input records  
**When** processing runs  
**Then** valid records are persisted  
**And** invalid records are skipped  
**And** summary counts are logged  

---

### Scenario 2: Idempotent Upsert
**Given** repeated ingestion of the same learner/course data  
**When** processing runs multiple times  
**Then** duplicates are avoided  
**And** latest state is retained per upsert policy  

---

### Scenario 3: Retrieval by Learner
**Given** a learner identifier  
**When** queried  
**Then** stored progress records are returned accurately and consistently  

---

### Scenario 4: Fault Tolerance
**Given** malformed timestamps or invalid progress values  
**When** records are processed  
**Then** invalid rows are rejected/skipped  
**And** the batch continues without aborting  

---

## Implementation Notes
- Keep validation, transformation, and persistence as clearly separated layers  
- Use consistent canonical model across the system  
- Ensure upsert logic aligns with unique constraints from schema layer  
- Design processing to be restart-safe and replay-safe  

---

## Validation Checklist
- [ ] Validation rules implemented and tested  
- [ ] Transformation logic produces consistent canonical format  
- [ ] Upsert logic prevents duplicates and preserves latest state  
- [ ] Batch processing continues on per-record failure  
- [ ] Summary reporting outputs correct counts  
- [ ] Retrieval endpoint/service returns correct learner data  
- [ ] Logs are structured and free of sensitive information  
- [ ] Linting and formatting checks pass

### PR DESCRIPTION
**Pull request by sserurich**
### Title
IC-13 feat: implement LearnWorlds transform and persistence pipeline

---

## Summary
Implements IC-13 for LearnWorlds data processing after ingestion by adding:

- A transformation and persistence pipeline that reads staged raw payloads and upserts learner progress records  
- An admin-only transform API endpoint  
- End-to-end Playwright coverage written 
---

## Related Issue
Closes IC-13

---

## Dependency / Context
This PR builds on the LearnWorlds ingestion flow introduced in IC-11.

IC-11 is responsible for:
- authenticating with LearnWorlds  
- fetching learner progress/activity data  
- storing sync runs and raw payloads  

IC-13 is responsible for:
- reading staged raw payloads  
- validating and transforming records  
- upserting normalized learner progress data  

Recommended review/merge order:
1. IC-11 — ingestion client + admin trigger endpoint  
2. IC-13 — transform and persistence pipeline  

---

### Tests (TDD)
- Added Playwright suite for `POST /api/admin/learnworlds/transform`  
- Added test utility to seed and clean LearnWorlds sync runs and raw payload records for integration scenarios  

**Coverage includes:**
- Unauthenticated request redirects to login  
- Authenticated non-admin returns `401`  
- Missing `syncRunId` returns `400`  
- Nonexistent `syncRunId` returns `404`  
- Mixed valid/invalid raw payloads are processed with expected persisted/skipped counts  
- Idempotency: running transform twice keeps stable counts via upsert behavior  

---

#### Transform Pipeline
- Loads staged raw payloads by `syncRunId`  
- Validates required identifiers (learner and course)  
- Skips invalid rows  
- Upserts valid rows into `learner_progress` using conflict target (`learnerId`, `courseId`)  
- Returns processed/persisted/skipped summary counts  

#### Admin Endpoint
- `POST /api/admin/learnworlds/transform`  
- Enforces admin authorization  
- Validates request body (`syncRunId`)  
- Verifies sync run exists  
- Returns structured success/error responses  

---

## Validation
- Playwright tests pass locally  
- Transform logic behaves deterministically across reruns  
- Idempotent upsert verified (no duplicate records)  
- Lint/type checks pass  

---

## Notes
- Builds on existing ingestion schema and raw payload staging  
- Strict separation maintained between raw payloads and normalized model  
- Pipeline is safe for reprocessing and partial failures  

---

### Functionality
- [ ] Transform pipeline processes raw payloads correctly  
- [ ] Valid records are persisted, invalid records are skipped  
- [ ] Summary counts (processed/persisted/skipped) are accurate  

### Idempotency & Data Integrity
- [ ] Upsert prevents duplicates (`learnerId`, `courseId`)  
- [ ] Re-running transform produces stable results  

### API Behavior
- [ ] Admin endpoint enforces authorization  
- [ ] Validation errors return correct status codes (`400`, `404`, `401`)  
- [ ] Response format is consistent  

### Testing
- [ ] Playwright tests pass  
- [ ] Edge cases (invalid data, missing syncRunId) are covered  

### Code Quality
- [ ] Code follows project conventions  
- [ ] No TypeScript errors  

### CI / Formatting
- [ ] Prettier passes
- [ ] Linting passes
- [ ] GitHub Actions checks are green

### CODE REVIEW

File: app/api/admin/learnworlds/transform/route.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import { db } from '@/lib/db';
+import { learnworldsSyncRuns } from '@/lib/db/schema';
+import { getUserFromSession } from '@/lib/auth/server';
+import { runLearnworldsTransform } from '@/lib/learnworlds/transform';
+
+interface TransformRequestBody {
+  syncRunId?: string;
+}
+
+function isLearnworldsSchemaUnavailableError(error: unknown): boolean {
```
Comment (8bit-Apollo) [inline app/api/admin/learnworlds/transform/route.ts:20]: A comment here explaining this function could enhance readability/clarity

File: tests/test-utils/seed-test-sync-run.ts [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+      learner_external_id: record.learnerExternalId,
+      course_external_id: record.courseExternalId,
+      completion_status: record.completionStatus,
+      progress_percentage: record.progressPercentage,
+      record_hash: hashPayload(raw),
+      payload: raw,
+    };
+  });
+
+  const payloadsRes = await fetch(`${SUPABASE_URL}/rest/v1/learnworlds_raw_payloads`, {
```
Comment (8bit-Apollo) [inline tests/test-utils/seed-test-sync-run.ts]: This should be fine for a small amount of records, but it might be better to do this in batches for the purpose of scalability to prevent large inserts from potentially timing out

**Review by 8bit-Apollo — COMMENTED** [review]
**tests/test-utils/seed-test-sync-run.ts** and **tests/test-utils/seed-test-sync-run.ts** could use more comments for the sake of clarity, but otherwise this is very good work. Code is fairly well organized and readable; testing is rigorous; and consistent without unnecessary redundancy.

**Comment by bofengb** [comment]
# PR Review: `feature/learnworlds-transform-persist`

Hey, nice work on this one. The pipeline separation between raw ingestion (#98) and transform/persist is clean, and I appreciate the idempotent upsert design. The Playwright test that runs the transform twice and asserts identical counts is a great way to prove that out. The FK-aware cleanup order in `seed-test-sync-run.ts` is well thought through too.

---

## 1. The Goal

This PR addresses IC-13 by adding a transform and persistence pipeline that reads staged raw payloads from #98's ingestion step and upserts normalized learner progress records. It includes an admin-only endpoint at `POST /api/admin/learnworlds/transform`, the core transform logic, comprehensive Playwright tests, and test seeding utilities. It builds on #98.

---

## 2. Issues

### 2.1 `completedModules` Is a Dead Column

In `lib/learnworlds/transform.ts` (line 45), `completedModules` is hardcoded to `0` on every insert:

```typescript
validPayloads.map((p) => ({
    learnerId: p.learnerExternalId!,
    courseId: p.courseExternalId!,
    progressPercentage: p.progressPercentage ?? 0,
    completedModules: 0,                              // ← always zero
    completionStatus: p.completionStatus ?? 'in_progress',
    lastActivityTimestamp: p.lastActivityTimestamp,
    rawPayloadId: p.id,
    sourceSyncedAt: new Date().toISOString(),
}))
```

And in the upsert's `set` clause (lines 51-56), `completedModules` is **not included**:

```typescript
.onConflictDoUpdate({
    target: [learnerProgress.learnerId, learnerProgress.courseId],
    set: {
        progressPercentage: sql`EXCLUDED.progress_percentage`,
        completionStatus: sql`EXCLUDED.completion_status`,
        lastActivityTimestamp: sql`EXCLUDED.last_activity_timestamp`,
        rawPayloadId: sql`EXCLUDED.raw_payload_id`,
        sourceSyncedAt: sql`EXCLUDED.source_synced_at`,
        // completedModules is absent
    },
});
```

This means:

1. On first insert, `completedModules` is set to `0`.
2. On subsequent upserts (conflict updates), `completedModules` is never touched. It's frozen at `0` forever.
3. Even if a future PR adds logic to compute `completedModules` on the initial insert, the upsert path would still skip it, so repeat transforms would silently regress the value back to whatever was set on first insert.

Meanwhile, the raw payloads from #98 store `moduleExternalId` and `lessonExternalId` for each record. The data needed to compute this field exists upstream, it's just not being used.

The `learner_progress` schema has this column, so downstream consumers (like the recommendation engine) might reasonably read it and assume it reflects reality. A column that's always `0` is worse than a column that doesn't exist. It looks like real data.

I'd suggest one of:

- **Compute it**: count distinct completed modules per learner+course from the raw payloads and set it on both insert and upsert. Add it to the `set` clause so repeat transforms update it correctly.
- **Defer it explicitly**: add a `// TODO(IC-XX)` comment explaining this is intentionally deferred, and consider whether the column should be nullable (with `null` meaning "not yet computed") rather than defaulting to `0`.

File: tests/test-utils/seed-test-sync-run.ts (hunk 2 above)
Comment (sserurich) [reply tests/test-utils/seed-test-sync-run.ts]: Thanks @8bit-Apollo  for the feedback. I have addressed both items:
- Added clearer comments/docs around schema-availability/error-handling and test utility flow.
- Refactored raw payload seeding to batch inserts for better scalability and lower timeout risk.
- Kept FK-safe cleanup order and documented it more explicitly.
- Verified with lint + Playwright transform coverage (all passing).

**Comment by sserurich** [comment]
> # PR Review: `feature/learnworlds-transform-persist`
> Hey, nice work on this one. The pipeline separation between raw ingestion (#98) and transform/persist is clean, and I appreciate the idempotent upsert design. The Playwright test that runs the transform twice and asserts identical counts is a great way to prove that out. The FK-aware cleanup order in `seed-test-sync-run.ts` is well thought through too.
> 
> ## 1. The Goal
> This PR addresses IC-13 by adding a transform and persistence pipeline that reads staged raw payloads from #98's ingestion step and upserts normalized learner progress records. It includes an admin-only endpoint at `POST /api/admin/learnworlds/transform`, the core transform logic, comprehensive Playwright tests, and test seeding utilities. It builds on #98.
> 
> ## 2. Issues
> ### 2.1 `completedModules` Is a Dead Column
> In `lib/learnworlds/transform.ts` (line 45), `completedModules` is hardcoded to `0` on every insert:
> 
> ```ts
> validPayloads.map((p) => ({
>     learnerId: p.learnerExternalId!,
>     courseId: p.courseExternalId!,
>     progressPercentage: p.progressPercentage ?? 0,
>     completedModules: 0,                              // ← always zero
>     completionStatus: p.completionStatus ?? 'in_progress',
>     lastActivityTimestamp: p.lastActivityTimestamp,
>     rawPayloadId: p.id,
>     sourceSyncedAt: new Date().toISOString(),
> }))
> ```
> 
> And in the upsert's `set` clause (lines 51-56), `completedModules` is **not included**:
> 
> ```ts
> .onConflictDoUpdate({
>     target: [learnerProgress.learnerId, learnerProgress.courseId],
>     set: {
>         progressPercentage: sql`EXCLUDED.progress_percentage`,
>         completionStatus: sql`EXCLUDED.completion_status`,
>         lastActivityTimestamp: sql`EXCLUDED.last_activity_timestamp`,
>         rawPayloadId: sql`EXCLUDED.raw_payload_id`,
>         sourceSyncedAt: sql`EXCLUDED.source_synced_at`,
>         // completedModules is absent
>     },
> });
> ```
> 
> This means:
> 
> 1. On first insert, `completedModules` is set to `0`.
> 2. On subsequent upserts (conflict updates), `completedModules` is never touched. It's frozen at `0` forever.
> 3. Even if a future PR adds logic to compute `completedModules` on the initial insert, the upsert path would still skip it, so repeat transforms would silently regress the value back to whatever was set on first insert.
> 
> Meanwhile, the raw payloads from #98 store `moduleExternalId` and `lessonExternalId` for each record. The data needed to compute this field exists upstream, it's just not being used.
> 
> The `learner_progress` schema has this column, so downstream consumers (like the recommendation engine) might reasonably read it and assume it reflects reality. A column that's always `0` is worse than a column that doesn't exist. It looks like real data.
> 
> I'd suggest one of:
> 
> * **Compute it**: count distinct completed modules per learner+course from the raw payloads and set it on both insert and upsert. Add it to the `set` clause so repeat transforms update it correctly.
> * **Defer it explicitly**: add a `// TODO(IC-XX)` comment explaining this is intentionally deferred, and consider whether the column should be nullable (with `null` meaning "not yet computed") rather than defaulting to `0`.

Thanks  @bofengb , great catch on `completedModules`. Implemented and validated the full fix:

- `completedModules` is now computed from distinct module IDs with `completionStatus = completed`.
- Upsert now reliably updates `completedModules` (and related fields) via `EXCLUDED` values.
- Aligned transform persistence with `learnworlds_user_id` naming.
- Hardened transform to aggregate to one row per user + course before upsert.
- Added integration tests that assert persisted DB values (status-aware counting, dedupe, null-module behavior, upsert updates, and multi-course same-user cases).

**Review by zsin520 — APPROVED** [review]
Hi Richard,

Looks good this is ready to merge.
